### ISSUE TITLE
Container Liquid Unloading feels Unintuitive

### ISSUE THREAD

**Issue by TheWebfrog** [CleverRaven/Cataclysm-DDA#20120]
This Issue is a reopen of #20115, which I spectacularly failed in documenting the issue.

**The issue:**

* Take a container filled with a liquid or solid, "water" in a plastic bottle in this example, and attempt to craft "clean water". 

* At this point make sure to only have the plastic bottle of water in your inventory.

* You will be prompted with "No Suitable Container to contain liquid". 

* At this point you cannot put the output into the input container; you cannot put the output "clean water" into the input container which contained "water".

* Aquire another plastic bottle or watertight container which is EMPTY. 

* Now attempt to craft "clean water" from the plastic bottle filled with "water". 

* Now not only will you find that you can pour "clean water" into the EMPTY empty plastic bottle, but you can also pour it into the original input container previously filled with "water".

This system feels clunky and unintuitive as if you had only the container with the input, you cannot put the result into it, but when you aquire another watertight container, you can put the result back into the originally unusable input container.

* Note only works with recipes in which the output is liquid.

* Conclusion

This issue could be fixed entirely by using the same tag which stops the tool being used in the recipe, but this tag is obscure and my cause problems.

* I am not sure what this tag is, but it can be easily found in action by trying to craft a recipe in which an item found in the recipe has the quality of tool needed in the recipe.

**Comment by keyspace** [comment]
Seems more crafting-related than unloading-related based on description.

Duplicate of #171.

**Comment by CoroNaut** [comment]
You technically wouldn't be using the container you pour what you made into in the crafting recipe. It's about taking the liquid out of the bottle, crafting with it, then filling it up again.

If the crafting system allowed a system to just take all of the required materials, craft with them, and then destroy them, it would solve the issue.

**Comment by keyspace** [comment]
> If the crafting system allowed a system to just take all of the required materials, craft with them, and then destroy them, it would solve the issue.

This approach could end up being multi-pass. I've not taken a look at how it's done now, but I'd guess it is already.

"Just take all" is not trivial. Consider crafting a molotov cocktail when there's a glass bottle of gasoline and a rag in inventory. The order in which components are collected into a "crafting buffer" matters.

* If the bottle is taken first, then the gasoline must be placed somewhere else. Crafting code could check-ahead if gasoline is perhaps required in the recipe - but if there's too much of it, the excess must be placed somewhere - bringing us back to square one.
* If the gasoline is taken first, and the bottle has some excess, should crafting code still consider the bottle, even if it's not empty?

All of this has to be done:

* _very fast_, so there's no lag when opening the crafting menu, or browsing recipes;
* with _no prompts_ - a hundred "maybe craftable" recipes would mean a hundred prompts.

This is a different example than OP, in that it uses no tools - the point is, whatever the algorithm, it _must_ work for all possible recipes.

-----

By the way! @TheWebfrog, which version is this?

Crafting clean water requires a tool with `boiling` quality. These tools often have `containing`, too, with the exception of some - e.g. coffeemaker, which is an unrelated bug. So one can craft clean water just fine, and pour it into the container that's also the used tool. Granted, these tools can be used for crafting even if filled, but that's another unrelated bug.

Anyway, the presented example can't be reproduced (on a recent git master).

> This issue could be fixed entirely by using the same tag which stops the tool being used in the recipe, but this tag is obscure and my cause problems.

Could you rephrase this?..

> I am not sure what this tag is, but it can be easily found in action by trying to craft a recipe in which an item found in the recipe has the quality of tool needed in the recipe.

Currently bugged, too: #18710

**Comment by Asmageddon** [comment]
What about a "craft anyway" option, for when you want to immediately consume it, or put it in a vehicle tank?

**Comment by Asmageddon** [comment]
@Leland: I do think that it has much in common, but isn't quite the same thing. Maybe a new issue for the "craft anyway" suggestion, since it solves both being able to use an empty tank, and the source container as target?
